### PR TITLE
add TemplateRepo, deprecate Template+Experiment

### LINKED ISSUE #331: GitHub 模板仓库接口
## 参数

GitHub 模板仓库 HTTP URL，如 https://github.com/idea2app/Next-Bootstrap-ts

## 逻辑

存储上述 URL 的同时，通过 GitHub API 读取并存储该仓库的：
- 语言：https://docs.github.com/cn/rest/repos/repos#list-repository-languages
- 标签：https://docs.github.com/cn/rest/repos/repos#get-all-repository-topics

### PR DESCRIPTION
**Pull request by hsupu**
close #331 

Notes:

- Add DI and mock for HttpClient

### CODE REVIEW

File: src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+                return (await JObject.LoadAsync(jsonReader, cancellationToken));
+            }
+        }
+
+        private async Task FetchGitHubInfo(TemplateRepo templateRepo, CancellationToken cancellationToken)
+        {
+            // Normalize the url.
+            var match = regexGitHubRepoUrl.Match(templateRepo.url);
+            if (!match.Success) {
+                return;
```
Comment (juniwang) [inline src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs:76]: Better to have logs here

File: src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            var owner = match.Groups[2].Value;
+            var repo = match.Groups[3].Value;
+
+            if (templateRepo.isFetched == true) {
+                return;
+            }
+
+            try {
+                // https://docs.github.com/en/rest/repos/repos#list-repository-languages
+                string url = string.Format("https://api.github.com/repos/{0}/{1}/languages", owner, repo);
```
Comment (juniwang) [inline src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs:88]: does this API allow anonymous call? If not, what's the required scope and how does OPH get it?

Comment (hsupu) [reply src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs:88]: Yes, I tested the API from local without cookie/token and it worked.

File: src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs (hunk 1 above)
Comment (hsupu) [reply src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs:77]: Added, but I don't know which one is better - `LogWarning` or `TraceInformation`.

File: src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs (hunk 2 above)
Comment (hsupu) [reply src/open-hackathon-server/Kaiyuanshe.OpenHackathon.Server/Biz/TemplateRepoManagement.cs:89]: I was wrong. GitHub API allows to access from browser, but it rejects when the HTTP request carrys minimal headers. We need a secret store first.

**Pull request merged**